operator/pkg/catalog/remote_catalog.go

```diff
@@ -0,0 +1,257 @@
+// as we already know the spec will change. It also saves us the use of any special status field.
+// If for some reason the generation is incremented multiple times in between the current and the next reconciliation
+// it can simply jump multiple generations by always basing it on the latest generation of the remote.
+func (*RemoteCatalog) CalculateDiffs(
+	runtimeList *v1alpha1.ModuleTemplateList, controlPlaneList *v1alpha1.ModuleTemplateList,
+) ([]*v1alpha1.ModuleTemplate, []*v1alpha1.ModuleTemplate) {
+	// these are various ModuleTemplate references which we will either have to create, update or delete from
+	// the remote
+	var diffToApply []*v1alpha1.ModuleTemplate
+	var diffToDelete []*v1alpha1.ModuleTemplate
+
+	// now lets start using two frequency maps to discover diffs
+	existingOnRemote := make(map[string]int)
+	existingOnControlPlane := make(map[string]int)
+
+	for i := range runtimeList.Items {
+		remote := &runtimeList.Items[i]
+		existingOnRemote[remote.Namespace+remote.Name] = i
+	}
+
+	for i := range controlPlaneList.Items {
+		controlPlane := &controlPlaneList.Items[i]
+		existingOnControlPlane[controlPlane.Namespace+controlPlane.Name] = i
+		// if the controlPlane Template does not exist in the remote, we already know we need to create it
+		// in the runtime
+		if _, exists := existingOnRemote[controlPlane.Namespace+controlPlane.Name]; !exists {
+			prepareControlPlaneTemplateForRuntime(controlPlane)
+
+			diffToApply = append(diffToApply, controlPlane)
+		}
+	}
+
+	for i := range runtimeList.Items {
+		remote := &runtimeList.Items[i]
+		controlPlaneIndex, exists := existingOnControlPlane[remote.Namespace+remote.Name]
+
+		// if the remote Template does not exist in the control plane, we already know we need to delete it
+		if !exists {
+			diffToDelete = append(diffToDelete, remote)
+			continue
+		}
+		remote.ObjectMeta.SetManagedFields([]metav1.ManagedFieldsEntry{})
+
+		if remote.Annotations == nil {
+			remote.Annotations = make(map[string]string)
+		}
+
+		// if there is a template in controlPlane and remote, but the generation is outdated, we need to
+		// update it
+		controlPlaneSyncedGen, _ := strconv.Atoi(remote.Annotations[v1alpha1.LastSyncGenerationControlPlane])
+		controlPlaneGen := (&controlPlaneList.Items[controlPlaneIndex]).GetGeneration()
+		runtimeSyncedGen, _ := strconv.Atoi(remote.Annotations[v1alpha1.LastSyncGenerationRuntime])
+
+		if int64(controlPlaneSyncedGen) != controlPlaneGen {
+			remote.Annotations[v1alpha1.LastSyncGenerationControlPlane] = strconv.Itoa(int(controlPlaneGen))
+			(&controlPlaneList.Items[controlPlaneIndex]).Spec.DeepCopyInto(&remote.Spec)
+			diffToApply = append(diffToApply, remote)
+		} else if int64(runtimeSyncedGen) != remote.GetGeneration() {
+			(&controlPlaneList.Items[controlPlaneIndex]).Spec.DeepCopyInto(&remote.Spec)
+			remote.Annotations[v1alpha1.LastSyncGenerationRuntime] = strconv.Itoa(int(remote.GetGeneration() + 1))
```

Comment:
better add a comment to explain this `+1`

Reply:
Its in the comment of the main method https://github.com/kyma-project/lifecycle-manager/pull/276/files#diff-b247ded7d56054153c2091874b28f2131d1b537bc921a7cbb7c6a82354f6b16aR106-R113